nucypher/network/retrieval.py

```diff
@@ -0,0 +1,410 @@
+        otherwise block and wait for them to come online.
+        """
+
+        # OK, so we're going to need to do some network activity for this retrieval.
+        # Let's make sure we've seeded.
+        if not self._learner.done_seeding:
+            self._learner.learn_from_teacher_node()
+
+        ursulas_in_map = treasure_map.destinations.keys()
+        all_known_ursulas = self._learner.known_nodes.addresses()
+
+        # Push all unknown Ursulas from the map in the queue for learning
+        unknown_ursulas = ursulas_in_map - all_known_ursulas
+        if unknown_ursulas:
+            self._learner.learn_about_specific_nodes(unknown_ursulas)
+
+        # How many nodes over the threshold we want to know (just in case)
+        redundancy = 0
+
+        # If we know enough to decrypt, we can proceed.
+        known_ursulas = ursulas_in_map & all_known_ursulas
+
+        if len(known_ursulas) - redundancy >= treasure_map.threshold:
+            return
+
+        allow_missing = max(len(unknown_ursulas) - redundancy, 0)
+        self._learner.block_until_specific_nodes_are_known(unknown_ursulas,
+                                                           timeout=timeout,
+                                                           allow_missing=allow_missing,
+                                                           learn_on_this_thread=True)
+
+    def _request_reencryption(self,
+                              ursula: 'Ursula',
+                              reencryption_request: 'ReencryptionRequest',
+                              delegating_key: PublicKey,
+                              receiving_key: PublicKey,
+                              ) -> Dict['Capsule', 'VerifiedCapsuleFrag']:
+        """
+        Sends a reencryption request to a single Ursula and processes the results.
+
+        Returns reencrypted capsule frags matched to corresponding capsules.
+        """
+
+        middleware = self._learner.network_middleware
+
+        try:
+            response = middleware.reencrypt(ursula, bytes(reencryption_request))
+        except NodeSeemsToBeDown as e:
+            # TODO: What to do here?  Ursula isn't supposed to be down.  NRN
+            message = (f"Ursula ({ursula}) seems to be down "
+                       f"while trying to complete ReencryptionRequest: {reencryption_request}")
+            self.log.info(message)
+            raise RuntimeError(message) from e
+        except middleware.NotFound as e:
+            # This Ursula claims not to have a matching KFrag.  Maybe this has been revoked?
+            # TODO: What's the thing to do here?
+            # Do we want to track these Ursulas in some way in case they're lying?  #567
+            message = (f"Ursula ({ursula}) claims not to have the KFrag "
+                       f"to complete ReencryptionRequest: {reencryption_request}. "
+                       f"Has access been revoked?")
```

Comment:
Since we are now sending the kfrags as part of the treasure map in the reencryption request this wording may need to be updated...?

Reply:
Yes, now we get `NotFound` if policy HRAC is not found. Adjusted the message.